CLDR-17754: added Samogitian language

zordsdavini opened the pull request:
CLDR-17754

- [x] This PR completes the ticket.

<!--
Thank you for your pull request.
Please see https://cldr.unicode.org/index/process for general
information on contributing to CLDR.

1. Make sure the ticket is filed at
https://unicode-org.atlassian.net/projects/CLDR/
2. Update the PR title and first line of this
message to include the ticket ID (CLDR-_____)
3. You will be automatically asked to sign the contributors’
license before the PR is accepted.
- sign: https://cla-assistant.io/unicode-org/cldr
- license: https://www.unicode.org/copyright.html#License
-->

ALLOW_MANY_COMMITS=true

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/unicode-org/cldr?pullRequest=3818) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/unicode-org/cldr?pullRequest=3818) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/unicode-org/cldr?pullRequest=3818) it.</sub>

conradarcturus commented on common/supplemental/plurals.xml:
```diff
@@ -225,6 +225,13 @@ CLDR data files are interpreted according to the LDML specification (http://unic
             <pluralRule count="many">v != 0   @decimal 0.0~1.5, 10.0, 100.0, 1000.0, 10000.0, 100000.0, 1000000.0, …</pluralRule>
             <pluralRule count="other"> @integer 3~10, 13~19, 23, 103, 1003, …</pluralRule>
         </pluralRules>
+        <pluralRules locales="sgs">
```
Most languages don't require so many variations. Often it is not necessary to have this full plural explosion. However it may totally be a feature of Samogitian so its totally fine to go ahead with this change.

Can you show examples of why you'd have 5 plural classes? Linguistically it looks like Samogitian should only have 3: Singular (one), Dual (two), and Plural (other). Can you write out some minimal pairs to show the difference? [Example from Lithuanian](https://www.unicode.org/cldr/charts/45/supplemental/language_plural_rules.html#lt):
1 obuolys
2 obuoliai
0,5 obuolio
10 obuolių

zordsdavini replied:
Yes, it is similar to Lithuanian, except Samogitian has dual form:
1 vuobols
2 vuobolo
3 vuobolē
0,5 vuobolė
10 vuoboliū

Lithuanian has Singular and Plural, but some quantities wants different inflection: vuobols, vuobolo, vuobolē (similar in Lithuanian obuolys, obuoliai) - all are in nominative and vuobolė (lt. obuolio) is singular genetive, vuololiū (lt. obuolių) is plural genetive

conradarcturus reviewed (changes requested):
It's really exciting to see this language added to CLDR :D Instead of adding everything all at once -- let's do it incrementally. See this pull request where I'm just adding the core data for Haitian Creole: https://github.com/unicode-org/cldr/pull/3948 -- Can you trim off most of the changes to `sgs.xml` in this diff to just the code data?

Once it's in the system we can contribute the time values in the Survey tool in the v47 window later this fall -- that tool has lots of assistance including validation.

Collation and casing and probably fine to add in this diff. I'm not so familiar w/ that code myself but it should generally be fine.

Also there are some other validation errors like that the pull request & commit should start w/ your Jira Ticket number. The Jira ticket does need to get reviewed but I tagged it so it should get review now.

zordsdavini commented on common/main/sgs.xml:
```diff
@@ -0,0 +1,1544 @@
+		<alternateQuotationStart>„</alternateQuotationStart>
+		<alternateQuotationEnd>“</alternateQuotationEnd>
+	</delimiters>
+	<dates>
```
removed a lot of. Let's try to start :) Added ticket numbers. Actually there are two tickets: https://unicode-org.atlassian.net/browse/CLDR-17754 and https://unicode-org.atlassian.net/browse/CLDR-17786

zordsdavini commented:
I checked errors and have problem
```
/home/runner/work/cldr/cldr/common/main/sgs.xml:18:110: error
sgs [Samogitian]	error	▸Core_Data|Alphabetic_Information|Characters_in_Use|Main_Letters◂	〈[a b c d e f g h i j k l m n o p q r s t u v w x y z]〉	【】	〈[aā b c č d eē ėė̄ f g h iī j k l m n oō p r s š t uū v z ž]〉	«=»	【】	⁅illegal characters in exemplars⁆	❮Error: Should be limited to (specific-script - uppercase - invisibles + İ); thus not contain: ‎[\u0304]‎❯	https://st.unicode.org/cldr-apps/v#/sgs//7ed88347aa1b55ed
Error:  Error: Should be limited to (specific-script - uppercase - invisibles + İ); thus not contain: ‎[\u0304]‎
```
We have letter "ė̄" (ė with macron or e with dot above and macron), ex.: gerė̄jē. Mostly we write it with ė and adding macron (\u0304). Unicode doesn't have such single letter and didn't want to add when we asked to do that. What should we do here? 
And we have keylayout with macron as separate key, so all ā, ē, ī, ō, ė̄, ū we can write in combination with \u0304

zordsdavini commented:
Should I add something more or do something because of failed build? The error looks very unrelated but maybe I am wrong?

DavidLRowe commented:
@srl295 Can you advise on test failures being reported here?

zordsdavini commented:
@srl295 is this branch merged or just deleted?

srl295 commented:
@zordsdavini Did not see this one (i thought only @conradarcturus was using that branch)

can you fix your commit so that it doesn't have the colon in it? 

If you click [Details](https://us-central1-dev-infra-273822.cloudfunctions.net/unicode-github-bot/info/unicode-org/cldr/3818) you can choose the Squash option there.

zordsdavini commented:
@srl295 tried to fix. I hope I did it correctly

srl295 commented on common/main/sgs.xml:
```diff
@@ -0,0 +1,36 @@
+	<identity>
+		<version number="$Revision$"/>
+		<language type="sgs"/>
+        <script type="Latin"/>
+		<territory type="LT"/>
```
```suggestion
```

drop these from this file

zordsdavini replied:
@srl295 but why? Should it be in other way said or this information is needless?

srl295 replied:
I missed this question - but it's incorrect in this file. This file is `sgs`.xml so it has no script or territory.

zordsdavini commented:
@conradarcturus should I do something more to make this move forward?

zordsdavini commented on common/main/sgs.xml:
```diff
@@ -0,0 +1,36 @@
+		<exemplarCharacters type="punctuation">[\- ‐‑ – — , ; \: ! ? . … “„ ( ) \[ \] \{ \}]</exemplarCharacters>
+	</characters>
+	<delimiters>
+		<quotationStart>„</quotationStart>
+		<quotationEnd>“</quotationEnd>
+		<alternateQuotationStart>„</alternateQuotationStart>
+		<alternateQuotationEnd>“</alternateQuotationEnd>
```
this follows same rules as Lithuanian, so that's not draft

srl295 replied:
maybe, but it should be sorted out in the survey tool.  You added data beyond the core requirements, so it hasn't been through the process yet.

srl295 replied:
core data is only these: https://cldr.unicode.org/index/cldr-spec/core-data-for-new-locales

(casing and collation are fine as those are separate files)

srl295 commented:
Needs tests to pass, then we can merge.

zordsdavini commented:
@srl295 I didn't noted your comment. Tried to understand the problem and fixed. At the beginning I followed LT rules, but it looks a little bit messy here, so I rewrote SGS part. Now it gives some conflict and needs to be reviewed

srl295 commented:
@zordsdavini you got bit by something unrelated.  Are you able to merge in the latest from CLDR main?  something like:

- `git fetch upstream`
- `git merge -m 'CLDR-17754 merge from main' upstream/main`

srl295 commented:
@zordsdavini possibly.

the title of the merge commit needed to start with `CLDR-17754` or the tooling will complain

srl295 commented:
@conradarcturus do you still have issues or can you approve?

srl295 commented:
> @srl295 this merge brought more issues. It should be rebased instead merged – now it complains about something in the middle. Second - tests with plural rules is failing. I don't understand why it gives zero. Here it is log:
> 
> ```
> Error:  (TestSupplementalInfo.java:743)  Error: : sgs, 2, one: expected false, got true
> 
> Error:  (TestSupplementalInfo.java:743)  Error: : sgs, 3, one: expected false, got true
> 
> Error:  (TestSupplementalInfo.java:743)  Error: : sgs, 4, one: expected false, got true
> 
> Error:  (TestSupplementalInfo.java:743)  Error: : sgs, 2, two: expected true, got false
> 
> Error:  (TestSupplementalInfo.java:743)  Error: : sgs, 3, two: expected true, got false
> 
> Error:  (TestSupplementalInfo.java:743)  Error: : sgs, 4, two: expected true, got false
> To fix the above, review the following, then replace in TestDigitPluralCompleteness
> {"sgs", "one", "0,00,000,0000"},	// n % 10 = 1 and n % 100 != 11
> {"sgs", "two", "0"},	// n = 2
> ```

I think they are only relevant issues. I'll take a look

conradarcturus commented:
Oh darn, you hit the dreaded exemption that the there is are new exemplar characters.

Do you have the CLDR environment set up? You can run the command below to rebuild the exemplar characters.

```
mvn -DSHOW_EXEMPLARS test --file tools/pom.xml '-Dorg.unicode.cldr.unittest.testArgs=-f:LikelySubtagsTest -n -q'
```

If you don't, I can make a separate pull request to add the `sgs.xml` -- then you can rebase this change on that for your plural changes.

conradarcturus reviewed (changes requested):
Need to rebuild exemplars:

```
mvn -DSHOW_EXEMPLARS test --file tools/pom.xml '-Dorg.unicode.cldr.unittest.testArgs=-f:LikelySubtagsTest -n -q'
```

zordsdavini commented:
@conradarcturus 

> Need to rebuild exemplars:

Tried but got error:
```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.11.0:compile (default-compile) on project cldr-code: Fatal error compiling: error: release version 11 not supported -> [Help 1]
```

conradarcturus commented:
We're running into many random test issues -- thanks for your patience @zordsdavini . Since we're opening the Survey Tool I want to unblock the locale so I made a separate PR that should add the code data for Samogitian -- I realized there were a few other files that needed to be updated too. https://github.com/unicode-org/cldr/pull/4621

If that looks good, we can rebase your change after we land that so you can add the plural rules.

zordsdavini commented:
@conradarcturus yes, let's move on. Sometimes it is better in small steps but forward

conradarcturus commented:
I landed the previously mentioned PR. You should be able to rebase this one now. There may be minor merge conflicts -- Note I had some tab/space fixes.

zordsdavini commented:
@conradarcturus I've made rebase with new your updates

conradarcturus commented:
Great! There was a plural test with an issue so I wonder if we need to update another file.

However, the good news is that Samogitian will be ready to submit on the survey tool. Can you request a survey tool account and next week I can add an account for you to fill in more Samogitian details.

https://docs.google.com/forms/d/e/1FAIpQLSdzt4W_EzVPw1N2hjOCtVxaGEHA1gFewslp7Zwj3gjJjrScRA/viewform?usp=sf_link

srl295 commented:
```
Error:  (TestSupplementalInfo.java:743)  Error: : sgs, 2, one: expected false, got true

Error:  (TestSupplementalInfo.java:743)  Error: : sgs, 3, one: expected false, got true

Error:  (TestSupplementalInfo.java:743)  Error: : sgs, 4, one: expected false, got true

Error:  (TestSupplementalInfo.java:743)  Error: : sgs, 2, two: expected true, got false

Error:  (TestSupplementalInfo.java:743)  Error: : sgs, 3, two: expected true, got false

Error:  (TestSupplementalInfo.java:743)  Error: : sgs, 4, two: expected true, got false
To fix the above, review the following, then replace in TestDigitPluralCompleteness
{"sgs", "one", "0,00,000,0000"},	// n % 10 = 1 and n % 100 != 11
{"sgs", "two", "0"},	// n = 2

```

I don't understand the test on a brief reading to understand what the issue is here.  Probably either:
1. the plural rules need work. Compare to Lithuanian?
2. the test needs an exception as it suggested

or both…

zordsdavini commented:
It is same as Lithuanian except Samogitian has dual for number 2. In Lithuanian there is something wrong with "few" and "other". I think they are missmatched.
For errors are wrong except 2=2

2025 m. balandžio 18 d. 23:21:30 GMT+03:00, "Steven R. Loomis" ***@***.***> parašė:
>srl295 left a comment (unicode-org/cldr#3818)
>
>```
>Error:  (TestSupplementalInfo.java:743)  Error: : sgs, 2, one: expected false, got true
>
>Error:  (TestSupplementalInfo.java:743)  Error: : sgs, 3, one: expected false, got true
>
>Error:  (TestSupplementalInfo.java:743)  Error: : sgs, 4, one: expected false, got true
>
>Error:  (TestSupplementalInfo.java:743)  Error: : sgs, 2, two: expected true, got false
>
>Error:  (TestSupplementalInfo.java:743)  Error: : sgs, 3, two: expected true, got false
>
>Error:  (TestSupplementalInfo.java:743)  Error: : sgs, 4, two: expected true, got false
>To fix the above, review the following, then replace in TestDigitPluralCompleteness
>{"sgs", "one", "0,00,000,0000"},	// n % 10 = 1 and n % 100 != 11
>{"sgs", "two", "0"},	// n = 2
>
>```
>
>I don't understand the test on a brief reading to understand what the issue is here.  Probably either:
>1. the plural rules need work. Compare to Lithuanian?
>2. the test needs an exception as it suggested
>
>or both…
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/unicode-org/cldr/pull/3818#issuecomment-2816133037
>You are receiving this because you were mentioned.
>
>Message ID: ***@***.***>

zordsdavini commented:
I tried to fix plurals. One thing is what according to rules `many` is used for fractions, so switched with `other`. And added some rule because it was empty with just examples (in other languages there are such examples). Needs to run approve workflow to check - locally maven throws 
```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.11.0:compile (default-compile) on project cldr-code: Fatal error compiling: error: release version 11 not supported -> [Help 1]
```

zordsdavini commented:
another error: other can't have any conditions: `Caused by: java.lang.IllegalArgumentException: The keyword 'other' must have no constraints, just samples.` - removing and rebased with master, too

zordsdavini commented:
I don't understand error - `build` can't create archives for earlier versions:
```
Exception in thread "main" java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.unicode.cldr.tool.Main.invoke(Main.java:59)
	at org.unicode.cldr.tool.Main.main(Main.java:46)
Caused by: java.lang.RuntimeException: Total errors: 46
	at org.unicode.cldr.tool.CheckoutArchive.main(CheckoutArchive.java:85)
	... 6 more
```
I can't fix it. `check` has passed at last
